### PR TITLE
Use ImageProjectiveTransformV3 for TF >= 2.4.0

### PR DESCRIPTION
**Pull request by WindQAQ**
# Description

Support fill_value and nearest fill_mode with ImageProjectiveTransformV3.

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [x] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

Test new features.

### CODE REVIEW

File: tensorflow_addons/image/transform_ops.py [hunk 1]
```diff
@@ -37,6 +37,7 @@ def transform(
     transforms: TensorLike,
     interpolation: str = "NEAREST",
     fill_mode: str = "CONSTANT",
+    fill_value: TensorLike = 0.0,
```
Comment (WindQAQ) [inline tensorflow_addons/image/transform_ops.py:40]: I have no idea of the where should we put the new argument. Should we maintain backward compatibility so we have to put it in the end of args? @seanpmorgan WDYT?

**Comment by WindQAQ** [comment]
@seanpmorgan Hi Sean, this is only for TF2.4. Our next release will pin to both TF2.3 and TF2.4 or 2.4 only?

**Comment by seanpmorgan** [comment]
> @seanpmorgan Hi Sean, this is only for TF2.4. Our nightly version will pin to both TF2.3 and TF2.4, right?

Trying to expand our python compatibility to be 2.3+. The wheel will be built against TF2.4 so custom-ops will need TF2.4, but everything else would be great to keep compatibility. 

Raising a NotImplementedError for <=2.4 is an option

**Comment by bhack** [comment]
Tracking many TF core issues and PRs my opinion Is that currently TF has a quite strict policy to not backport not only features but also bug fixes from nightly/master. Patch releases are sparse and for security fixes only.

So I think that we need to start to support and promote only the last official TF release.

**Comment by WindQAQ** [comment]
> Trying to expand our python compatibility to be 2.3+. The wheel will be built against TF2.4 so custom-ops will need TF2.4, but everything else would be great to keep compatibility.
> 
> Raising a NotImplementedError for <=2.4 is an option

I can do a warning for TF<2.4 that `fill_value` is not supported and is always 0.

> Tracking many TF core issues and PRs my opinion Is that currently TF has a quite strict policy to not backport not only features but also bug fixes from nightly/master. Patch releases are sparse and for security fixes only.
> 
> So I think that we need to start to support and promote only the last official TF release.

I am more with this proposal though, there is always a loop from "I want the new feature" -> "Please use latest TF" -> "codes does not work in latest TF" -> "Fallback to older TF" -> "I want the new feature" XD

**Comment by bhack** [comment]
> I am more with this proposal though, there is always a loop from "I want the new feature" -> "Please use latest TF" -> "codes does not work in latest TF" -> "Fallback to older TF" -> "I want the new feature" XD

Often it Is more like:
There Is a bug -> It Is fixed in the last version, or in nightly/master -> please use the last official or nightly wheels.

File: tensorflow_addons/image/transform_ops.py [hunk 2]
```diff
@@ -120,17 +123,36 @@ def transform(
...
+                images=images,
+                transforms=transforms,
+                output_shape=output_shape,
+                interpolation=interpolation.upper(),
+                fill_mode=fill_mode.upper(),
+                fill_value=fill_value,
+            )
+        else:
+            fill_mode = fill_mode.upper()
+            # TODO(WindQAQ): Get rid of the check once we drop TensorFlow < 2.4 support.
```
Comment (bhack) [inline tensorflow_addons/image/transform_ops.py:140]: Can we automatically/programmatically fail with https://github.com/tensorflow/addons/blob/master/setup.py#L79?
What do you think? I suppose that this kind of things will stay forever in the code :wink:

File: tensorflow_addons/image/transform_ops.py [hunk 3]
```diff
@@ -37,6 +40,7 @@ def transform(
     transforms: TensorLike,
     interpolation: str = "NEAREST",
     fill_mode: str = "CONSTANT",
```
Comment (aaronmondal) [inline tensorflow_addons/image/transform_ops.py:42]: Could we set these set these strings to small caps `"nearest"` and `"constant"` (in the `Args` below as well)? This would be more consistent with the docstrings in `tf.keras.layers` and we have an `.upper()` in the code anyways.

File: tensorflow_addons/image/transform_ops.py [hunk 4]
```diff
@@ -57,14 +61,18 @@ def transform(
...
-        to the given mode (one of `{'constant', 'reflect', 'wrap'}`).
+        to the given mode (one of `{'constant', 'reflect', 'wrap', 'nearest'}`).
         - *reflect*: `(d c b a | a b c d | d c b a)`
           The input is extended by reflecting about the edge of the last pixel.
         - *constant*: `(k k k k | a b c d | k k k k)`
           The input is extended by filling all values beyond the edge with the
           same constant value k = 0.
         - *wrap*: `(a b c d | a b c d | a b c d)`
           The input is extended by wrapping around to the opposite edge.
+        - *nearest*: `(a a a a | a b c d | d d d d)`
```
Comment (aaronmondal) [inline tensorflow_addons/image/transform_ops.py:72]: Does this actually work? According to the docs for `ProjectiveTransformV3` the `'nearest'` option is for the `interpolation` arg, not the `fill_value` arg. There is no test for `nearest` in `transform_ops_test.py` either.

File: tensorflow_addons/image/translate_ops.py [hunk 5]
```diff
@@ -76,6 +76,7 @@ def translate(
     translations: TensorLike,
     interpolation: str = "NEAREST",
     fill_mode: str = "CONSTANT",
```
Comment (aaronmondal) [inline tensorflow_addons/image/translate_ops.py:78]: Same here.

File: tensorflow_addons/image/translate_ops.py [hunk 6]
```diff
@@ -92,14 +93,18 @@ def translate(
...
-        to the given mode (one of `{'constant', 'reflect', 'wrap'}`).
+        to the given mode (one of `{'constant', 'reflect', 'wrap', 'nearest'}`).
         - *reflect*: `(d c b a | a b c d | d c b a)`
           The input is extended by reflecting about the edge of the last pixel.
         - *constant*: `(k k k k | a b c d | k k k k)`
           The input is extended by filling all values beyond the edge with the
           same constant value k = 0.
         - *wrap*: `(a b c d | a b c d | a b c d)`
           The input is extended by wrapping around to the opposite edge.
+        - *nearest*: `(a a a a | a b c d | d d d d)`
```
Comment (aaronmondal) [inline tensorflow_addons/image/translate_ops.py:104]: And here.

**Review by aaronmondal — CHANGES_REQUESTED** [review]
Thanks I need this so much! Was going crazy over the inconsistency between the different fill options for geometric transformations :D.

The only thing I am not so sure about is whether the `nearest` fill option actually does work.

There should probably be tests for the `"nearest"` and `"bilinear"` **interpolation** modes of `transform` as well. Currently, the `interpolation` argument is untested.

The `test_bilinear` and `test_uint8` tests in `transform_ops_test.py` apply to `rotate` only, which does make use of `transform` internally, but I don't think this relation should be assumed by the test case.

File: tensorflow_addons/image/transform_ops.py (hunk 4 above)
Comment (WindQAQ) [reply tensorflow_addons/image/transform_ops.py:72]: Yes it works, I forgot to add cc documentation in old PR :-(

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/image/image_ops.h#L113-L119

**Comment by fsx950223** [comment]
The function should be transfered to official repo? There is a similar function https://github.com/tensorflow/tensorflow/blob/2a7d45c6800a99e4158526e25ec128ddd90e37f7/tensorflow/python/keras/layers/preprocessing/image_preprocessing.py#L625.

**Comment by WindQAQ** [comment]
> The function should be transfered to official repo? There is a similar function https://github.com/tensorflow/tensorflow/blob/2a7d45c6800a99e4158526e25ec128ddd90e37f7/tensorflow/python/keras/layers/preprocessing/image_preprocessing.py#L625.

They do not make it public. In fact, they migrate it to KPL without informing us early this year 😭

**Comment by bhack** [comment]
> 
> They do not make it public. In fact, they migrate it to KPL without informing us early this year sob So we are just doing a fork of private API now.

Lets see :wink: ... https://github.com/tensorflow/tensorflow/pull/45742

File: tensorflow_addons/image/transform_ops.py (hunk 1 above)
Comment (seanpmorgan) [reply tensorflow_addons/image/transform_ops.py:43]: Yes this looks like the best position

File: tensorflow_addons/image/transform_ops.py (hunk 2 above)
Comment (seanpmorgan) [reply tensorflow_addons/image/transform_ops.py:140]: Eh, we can clean up the codebase as our compatibility matrix moves. See https://github.com/tensorflow/addons/pull/2294

**Comment by seanpmorgan** [comment]
> > They do not make it public. In fact, they migrate it to KPL without informing us early this year sob So we are just doing a fork of private API now.
> 
> Lets see 😉 ... [tensorflow/tensorflow#45742](https://github.com/tensorflow/tensorflow/pull/45742)

Agree with the PR, though I don't want to have to wait for a response prior to 0.12 release.

**Comment by bhack** [comment]
> Agree with the PR, though I don't want to have to wait for a response prior to 0.12 release.

It is impossible to achieve this for 0.12 cause this will got in TF 2.5 but you can bypass the policy and depend on a private API for this case, as you want.

**Comment by seanpmorgan** [comment]
> > Agree with the PR, though I don't want to have to wait for a response prior to 0.12 release.
> 
> It is impossible to achieve this for 0.12 cause this will got in TF 2.5 but you can bypass the policy and depend on a private API for this case, as you want.

Understand the TF core PR wouldn't be until TF2.5, but what do you mean depend on a private API and bypass the policy? There is no private API in this PR right?

**Comment by bhack** [comment]
> > > Agree with the PR, though I don't want to have to wait for a response prior to 0.12 release.
> > 
> > 
> > It is impossible to achieve this for 0.12 cause this will got in TF 2.5 but you can bypass the policy and depend on a private API for this case, as you want.
> 
> Understand the TF core PR wouldn't be until TF2.5, but what do you mean depend on a private API and bypass the policy? There is no private API in this PR right?

I meant as an alternative to maintain a (function) fork.

File: tensorflow_addons/image/transform_ops.py (hunk 1 above)
Comment (WindQAQ) [reply tensorflow_addons/image/transform_ops.py]: Updated.

File: tensorflow_addons/image/transform_ops.py (hunk 3 above)
Comment (WindQAQ) [reply tensorflow_addons/image/transform_ops.py]: Updated.

File: tensorflow_addons/image/transform_ops.py [hunk 7]
```diff
@@ -35,10 +38,11 @@
 def transform(
     images: TensorLike,
     transforms: TensorLike,
-    interpolation: str = "NEAREST",
-    fill_mode: str = "CONSTANT",
+    interpolation: str = "nearest",
+    fill_mode: str = "constant",
```
Comment (WindQAQ) [inline tensorflow_addons/image/transform_ops.py:42]: This will not break the compatibility as we always uppercase it when passing into raw ops.

**Comment by aaronmondal** [comment]
@WindQAQ Thanks for changing to lowercase. So much prettier :). Would you like me to file a PR for the missing testcases in `interpolation`?

**Comment by WindQAQ** [comment]
> @WindQAQ Thanks for changing to lowercase. So much prettier :). Would you like me to file a PR for the missing testcases in `interpolation`?

It's more than welcome to do that!

**Review by aaronmondal — APPROVED** [review]

**Review by seanpmorgan — APPROVED** [review]
LGTM Thanks! Thanks @aaronmondal for the review!

File: tensorflow_addons/image/transform_ops.py (hunk 2 above)
Comment (bhack) [reply tensorflow_addons/image/transform_ops.py:140]: Yes but It could be easier if we are failing in different points in the sourfe when we move the minimal version or the single version we support. Do you understand what I meant?

**Comment by bhack** [comment]
> > > > Agree with the PR, though I don't want to have to wait for a response prior to 0.12 release.
> > > 
> > > 
> > > It is impossible to achieve this for 0.12 cause this will got in TF 2.5 but you can bypass the policy and depend on a private API for this case, as you want.
> > 
> > 
> > Understand the TF core PR wouldn't be until TF2.5, but what do you mean depend on a private API and bypass the policy? There is no private API in this PR right?
> 
> I meant as an alternative to maintain a (function) fork.

It seems that we need to maintain the fork but if the new policy Is to maintain the public API surface minimal isn't better to handle private API occasional breakage instead of maintain fork in multiple points in the source code?
